openml/extensions/extension_interface.py

```diff
@@ -0,0 +1,291 @@
+from abc import ABC, abstractmethod
+from collections import OrderedDict  # noqa: F401
+from typing import Any, Dict, List, Optional, Tuple, TYPE_CHECKING
+
+
+if TYPE_CHECKING:
```

Comment:
I'll let Matthias correct me if I'm wrong, but short answer: this helps prevent cyclic imports.

Long answer: `TYPE_CHECKING` is a flag that is used to determine if a module is being loaded as part of a (static) type checking run. In order to do good type checking (e.g. proper use/availability of methods), normally you annotate the types directly (`can_handle_flow(flow : OpenMLFlow) -> bool`). However, to do this, you would have to (in this case) import `OpenMLFlow`, which itself would import (indirectly) again this file, leading to cyclic imports. Now you can still indicate the intended type by just specifying a string literal (`'OpenMLFlow'`), and while this helps for readability, the type checker itself does not know exactly what type it is, an import is still needed. So, using the `TYPE_CHECKING` flag the import can be done in type checking time but not at runtime (I am not sure about how type checkers avoid the cyclic imports exactly, but that's the worry for them (or anyone setting the `typing.TYPE_CHECKING` flag)). So, to answer the second question, `TYPE_CHECKING` should always be `False` unless a package (such as `mypy`) specifically sets it otherwise, in which case they are responsible for how things are handled (the file wouldn't normally load with `TYPE_CHECKING=True` due to the cyclic imports).

Reply:
Thanks @PGijsbers for adding the explanation, I believe it is correct.

I'll add a one-line comment to the if statement to make it easier to look this up.